python/cuml/common/array.py

```diff
@@ -168,7 +168,10 @@ def __getitem__(self, slice):
     def __setitem__(self, slice, value):
         cp.asarray(self).__setitem__(slice, value)
 
-    def __reduce__(self):
+    def __reduce_ex__(self, protocol):
+        return self.__class__, (self.to_output('numpy'),)
```

Comment:
Would suggest wrapping the NumPy array in a `PickleBuffer` when the pickle protocol used supports it.

```suggestion
-        return self.__class__, (self.to_output('numpy'),)
+        data = self.to_host_array()
+        if protocol >= 5:
+            data = pickle.PickleBuffer(data)
+        return self.__class__, (data,)
```

Reply:
@jakirkham doing that was the problem that broke cumlarray (inherited from cuDF buffer), since it makes it that at unpickling the data type is messed up, cumlarray is not meant no be a bytes/buffer object unlike cuDF.

Reply:
Ah nevermind, I totally misunderstood/misremembered the code in cuDF for this, by using our to_output as opposed to cuDF's to_host_array() the datatype would be kept and we keep the optimization, so nevermind my above comment you are right

Reply:
No worries. I guess Python seems to prefer `__reduce_ex__` even in subclasses that have implemented some other pickling method. So your change should fix that.

The idea here is to benefit from out-of-band buffer serialization when pickle is used.

Am hoping PR ( https://github.com/rapidsai/cudf/pull/5139 ) will simplify things by adding an ABC that fills in things like `__reduce_ex__` and only relies on `serialize`/`deserialize` methods to handle the underlying implementation details.